src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,569 @@
+        }
+
+        /// <summary>
+        /// Writes the start boundary for an operation. This is either the batch or the changeset boundary.
+        /// </summary>
+        protected override void WriteStartBoundaryForOperation()
+        {
+            if (this.changeSetBoundary == null)
+            {
+                ODataBatchWriterUtils.WriteStartBoundary(this.RawOutputContext.TextWriter, this.batchBoundary, !this.batchStartBoundaryWritten);
+                this.batchStartBoundaryWritten = true;
+            }
+            else
+            {
+                ODataBatchWriterUtils.WriteStartBoundary(this.RawOutputContext.TextWriter, this.changeSetBoundary, !this.changesetStartBoundaryWritten);
+                this.changesetStartBoundaryWritten = true;
+            }
+        }
+
+        /// <summary>
+        /// Remember a non-null Content-ID header for change set request operations.
+        /// If a non-null content ID header is specified for a change set request operation, record it in the URL resolver.
+        /// </summary>
+        /// <param name="contentId">The Content-ID header value read from the message.</param>
+        /// <remarks>
+        /// Note that the content ID of this operation will only
+        /// become visible once this operation has been written
+        /// and OperationCompleted has been called on the URL resolver.
+        /// </remarks>
+        private void RememberContentIdHeader(string contentId)
+        {
+            // The Content-ID header is only supported in request messages and inside of changeSets.
+            Debug.Assert(this.CurrentOperationRequestMessage != null, "this.CurrentOperationRequestMessage != null");
+            Debug.Assert(this.changeSetBoundary != null, "this.changeSetBoundary != null");
+
+            // Set the current content ID. If no Content-ID header is found in the message,
+            // the 'contentId' argument will be null and this will reset the current operation content ID field.
+            this.CurrentOperationContentId = contentId;
+
+            // Check for duplicate content IDs; we have to do this here instead of in the cache itself
+            // since the content ID of the last operation never gets added to the cache but we still
+            // want to fail on the duplicate.
+            if (contentId != null && this.PayloadUriConverter.ContainsContentId(contentId))
+            {
+                throw new ODataException(Strings.ODataBatchWriter_DuplicateContentIDsNotAllowed(contentId));
+            }
+        }
+
+        /// <summary>
+        /// Verifies that, for the case within a changeset, CreateOperationRequestMessage is valid.
+        /// </summary>
+        /// <param name="method">The HTTP method to be validated.</param>
+        /// <param name="contentId">The content Id string to be validated.</param>
+        private void VerifyCanCreateOperationRequestMessageAgainstChangeSetBoundary(string method, string contentId)

```

Comment:
>VerifyCanCreateOperationRequestMessageAgainstChangeSetBoundary [](start = 21, length = 62)

Put this check in CreateOperationRequestMessageImplementation and you can make VerifyCanCreateOperationRequestMessage private to ODataBatchWriter (the same as VerifyCanCreateOperationResponseMessage).

Reply:
Verifications should be completely done in Verify...() method. CreateOperationRequestMessageImplementation() should never perform verification which could throw verification errors. Right?

---
In reply to: [140936269](https://github.com/OData/odata.net/pull/960#discussion_r140936269) [](ancestors = 140936269)

Reply:
see above; we can move this logic into common method.

---
In reply to: [142586760](https://github.com/OData/odata.net/pull/960#discussion_r142586760) [](ancestors = 142586760,140936269)

Reply:
OK, as long as we keep the verification done completely. Changed this method to overriding base-class implementation (which is a no-op).

---
In reply to: [142821481](https://github.com/OData/odata.net/pull/960#discussion_r142821481) [](ancestors = 142821481,142586760,140936269)